src/extra/components/adq.py

```diff
@@ -0,0 +1,1294 @@
+
+        elif shape == 'pulses':
+            out = np.zeros(shape=pulse_shape, dtype=data.dtype)
+            self._reshape_flat_pulses(data, out, pulse_ids.to_numpy(),
+                                      out.shape[1])
+        else:
+            raise ValueError('shape must be one of `trains`, `pulses`')
+
+        return out
+
+    def find_edges(self, data,
+                   edge_func=None, max_edges=50, with_amplitudes=False,
+                   sample_dim='sample',  shape='edges', squeeze_edges=True,
+                   pasha_ctx=None, **edge_kw):
+        """Find signal edges by pulse.
+
+        In some cases, not the raw data itself may be of interest but
+        the location (and amplitude) of certain signals in the raw data.
+        One such example is time-of-flight spectroscopy in counting
+        mode, where individual charged particles impact a detection
+        surface and leave a fast signal on the digitizer trace. Fast
+        timing discriminators allow to robustly determine the position
+        of such signals.
+
+        By default, it uses the
+        [dynamic leading discriminator](extra.signal.dled) from the
+        [extra.signal](extra.signal) package, but other from this
+        package or entirely custom functions may be used as well. The
+        required signatures must include three keyword arguments
+        `signal`, `edges` and `amplitudes` corresponding to those from
+        [extra.signal.dled](extra.signal.dled). The default edge finding
+        method requires the `threshold` parameter to be passed as
+        keyword argument.
+
+        The processing is parallelized via `pasha`.
+
+        Args:
+            data (ArrayLike): Input data to find edges on.
+            edge_func (Callable, optional): Edge finding method to run
+                on each train trace, extra.signal.dled by default.
+            max_edges (int, optional): Maximal number of edges per
+                train, 1/5000 of trace length by default.
+            with_amplitudes (bool, optional): Full pulse height
+                corresponding to the determined edge positions.
+            sample_dim ({sample, time}, optional): Coordinates for
+                sample dimension if a labelled result is returned,
+                'sample' by default.
+            shape ({edges, input}, optional): Whether to keep a single
+                edge dimension (`edges`, default) or preserve the input
+                shape (`input`). The latter option returns labelled
+                xarray objects, while the former returns pandas
+                series or dataframe.
+            squeeze_edges (bool, optional): Whether to minimize the edge
+                axis lengh for xarray objects, True by default and
+                ignored when pandas objects are returned.
+            pasha_ctx (pasha.MapContext, optional): Pasha context to run
+                parallelized operations, pasha.ProcessContext by default.
```

Comment:
I'm always wary of parallel-by-default methods because it doesn't compose well - if you call this from inside code that's already parallelised, you're likely to get a mysterious error about 'daemon children'. And the magic incantation to get past that will be something like `pasha_ctx=pasha.SerialContext()`.

Can I suggest that instead of exposing the implementation, we make the option something like `parallel=N`, and have a special case for 0 or False to mean run in serial?

Reply:
Amusingly it was exactly like that before, a `num_workers` argument that switched to `SerialContext` if `<= 1`. 

Shortly before pushing I thought a `pasha` context would be more powerful. But I did not so much consider the fact you might want to disable it, but rather customize it more. But the disable case is likely stronger.

Reply:
I'd just like to point out that in my experience vanishingly few people 1) know how, and 2) actually regularly do multiprocessing in Python. I would be in favour of making it parallel by default for the 95% of people who don't do multiprocessing and give the others an informative error message.

Reply:
I went with Thomas' suggestion of `parallel=N`, though using `None` as default to express the `min(10, cpu_cores // 4)` logic. This should make it convenient, but transparent and easy to disable. In hindsight, there's no point (yet?) in doing thread-based parallelization except for `find_edges()`, I will evaluate that later.

Reply:
Thanks, that works for me. :slightly_smiling_face: 

In general I think methods that do parallel processing internally should know what kind of parallelism is likely to work, e.g. threads vs processes, and that shouldn't need to be a parameter. Admittedly that's hard if you can pass an arbitrary function in, but then I think passing functions into APIs is usually a kind of awkward design.

Reply:
Yes, I actually toned the arbitrary function part down already a notch. It is using the DLED now by default, which is very robust and should serve pretty much any discrimination needs. I left the arbitrary function part in, as the machinery could easily be used for any other form of peak finding a user might supply. It seemed cruel to force reimplementation just for API design sake.